x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -694,27 +696,52 @@ protected Expression rule(Expression e) {
 
         @Override
         protected Expression rule(Expression e) {
-            if (e instanceof ArbitraryConditionalFunction) {
-                ArbitraryConditionalFunction c = (ArbitraryConditionalFunction) e;
+            if (e instanceof ConditionalFunction) {
+                List<Expression> children = e.children();
+                // optimize nullIf
+                if (e instanceof NullIf) {
+                    NullIf nullIf = (NullIf) e;
+                    if (Expressions.isNull(nullIf.left()) || Expressions.isNull(nullIf.right())) {
+                        return nullIf.left();
+                    }
+                }
+                if (e instanceof ArbitraryConditionalFunction) {
+                    ArbitraryConditionalFunction c = (ArbitraryConditionalFunction) e;
+
+                    // there's no need for a conditional if all the children are the same (this includes the case of just one value)
+                    if (c instanceof Coalesce && children.size() > 0) {
+                        Expression firstChild = children.get(0);
+                        boolean sameChild = true;
+                        for (int i = 1; i < children.size(); i++) {
+                            Expression child =  children.get(i);
```

Comment:
Nit: there's an extra space after equals sign, but the assignment could also be done away with, `child`'s only used once in following expression.